node: reduce unsafe uint256S usage

stickies-v opened the pull request:
Since fad2991ba073de0bd1f12e42bf0fbaca4a265508, `uint256S` has been [deprecated](https://github.com/bitcoin/bitcoin/pull/30482/commits/fad2991ba073de0bd1f12e42bf0fbaca4a265508#diff-800776e2dda39116e889839f69409571a5d397de048a141da7e4003bc099e3e2R138) because it is less robust than the `base_blob::FromHex()` introduced in [the same PR](https://github.com/bitcoin/bitcoin/pull/30482). Specifically, it tries to recover from length-mismatches, recover from untrimmed whitespace, 0x-prefix and garbage at the end, instead of simply requiring exactly 64 hex-only characters. _(see also #30532)_

This PR carves out the few `uint256S` callsites that may potentially prove a bit more controversial to change because they deal with user input and backwards incompatible behaviour change.

The main behaviour change introduced in this PR is:
- `-minimumchainwork` will raise an error when input is longer than 64 hex digits
- `-assumevalid` will raise an error when input contains invalid hex characters, or when it is longer than 64 hex digits
- test: the optional RANDOM_CTX_SEED env var will now cause tests to abort when it contains invalid hex characters, or when it is longer than 64 hex digits

After this PR, the remaining work to remove `uint256S` completely is almost entirely mechanical and/or test related. I will open that PR once #30560 is merged because it builds on that.

stickies-v committed:
node: use uint256::FromUserHex for -minimumchainwork parsing

Removes dependency on unsafe and deprecated uint256S.

This makes parsing more strict, by returning an error
when the input contains more than 64 hex digits.

stickies-v committed:
node: use uint256::FromUserHex for -assumevalid parsing

Removes dependency on unsafe and deprecated uint256S.

This makes parsing more strict, by returning an error
when the input contains non-hex characters, or when it
contains more than 64 hex digits.

Also make feature_assumevalid.py more robust by using CBlock.hash
which is guaranteed to be 64 characters long, as opposed to the
variable-length hex(CBlock.sha256)

stickies-v committed:
test: use uint256::FromUserHex for RANDOM_CTX_SEED

Removes dependency on unsafe and deprecated uint256S.

This makes parsing more strict, by requiring RANDOM_CTX_SEED
to be a string of up to 64 hex digits (optionally prefixed with
"0x"), whereas previously any string would be accepted, with
non-hex characters silently ignored and input longer than
64 characters (ignoring "0x" prefix) silently trimmed.

Can be tested with:

$ RANDOM_CTX_SEED=z ./src/test/test_bitcoin --log_level=all --run_test=timeoffsets_tests/timeoffsets_warning -- -printtoconsole=1 | grep RANDOM_CTX_SEED
RANDOM_CTX_SEED must consist of up to 64 hex digits ("0x" prefix allowed), it was set to: 'z'.

Co-Authored-By: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

maflcko reviewed (changes requested):
Not sure about the breaking changes. `-noassumevalid` seems common https://github.com/search?q=noassumevalid&type=code

stickies-v commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -32,13 +32,20 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
+        if (auto block_hash{uint256::FromHex(util::RemovePrefixView(*value, "0x"))}) {
+            opts.assumed_valid_block = *block_hash;
+        } else {
+            return util::Error{strprintf(Untranslated("Invalid assumevalid block hash specified (%s), must be %d character hex"), *value, uint256::size() * 2)};
```
Ah thanks, didn't think of that. Will add a "0" check:

<details>
<summary>git diff on c0d508e69c</summary>

```diff
diff --git a/src/node/chainstatemanager_args.cpp b/src/node/chainstatemanager_args.cpp
index 8085aa65dc..60cee3dd78 100644
--- a/src/node/chainstatemanager_args.cpp
+++ b/src/node/chainstatemanager_args.cpp
@@ -31,7 +31,7 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
 
     if (auto value{args.GetBoolArg("-checkpoints")}) opts.checkpoints_enabled = *value;
 
-    if (auto value{args.GetArg("-minimumchainwork")}) {
+    if (auto value{args.GetArg("-minimumchainwork")}; value && *value != "0") {
         if (auto min_work{uint256::FromHex(util::RemovePrefixView(*value, "0x"))}) {
             opts.minimum_chain_work = UintToArith256(*min_work);
         } else {
@@ -39,7 +39,7 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
         }
     }
 
-    if (auto value{args.GetArg("-assumevalid")}) {
+    if (auto value{args.GetArg("-assumevalid")}; value && *value != "0") {
         if (auto block_hash{uint256::FromHex(util::RemovePrefixView(*value, "0x"))}) {
             opts.assumed_valid_block = *block_hash;
         } else {

```
</details>

stickies-v replied:
`-minimumchainwork` now accepts hex number input (including "0") again, and `assumevalid` has an extra `*value != "0"` check. Marking as resolved.

stickies-v commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -32,13 +32,20 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
+        if (auto min_work{uint256::FromHex(util::RemovePrefixView(*value, "0x"))}) {
+            opts.minimum_chain_work = UintToArith256(*min_work);
+        } else {
+            return util::Error{strprintf(Untranslated("Invalid minimum chain work value specified (%s), must be %d character hex"), *value, uint256::size() * 2)};
```
Thanks for the suggestion. I've been exploring different options here. Adding a `bool is_num` optional param to `uint256::FromHex()` is an option, but felt a bit awkward because it'd be a uint256-specific parameter (doesn't make sense for `transaction_identifier`).

Since we only need this for user input validation, I've now taken a different approach where I've repurposed `IsHexNumber()` -> `TrySanitizeHexNumber()`. I think this minimizes churn (e.g. the test suite can largely be kept as is), isolates user input validation from code that's mostly used internally and benefits from being strict and simple.

stickies-v commented on src/test/util/random.cpp:
```diff
@@ -9,24 +9,26 @@
+        if (const auto num{std::getenv(RANDOM_CTX_SEED)}) {
+            uint256 rv;
+            rv.SetHexDeprecated(num);
+            return rv;
```
Done (and RANDOM_CTX_SEED now also accepts < 64 character input string), thanks.

stickies-v replied:
> However, in the tests, personally I like to use it for brevity.

I agree. The tests fail gracefully, and the error message is helpful, including the location of the assertion which documents that we're expecting a 64 char hex string. Going to leave as is.

```
Running 1 test case...
test/util/random.cpp:29 operator(): Assertion `uint256::FromHex(num)' failed.
unknown location:0: fatal error: in "timeoffsets_tests/timeoffsets_warning": signal: SIGABRT (application abort requested)
test/timeoffsets_tests.cpp:62: last checkpoint: "timeoffsets_warning" fixture ctor

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
```

hodlinator replied:
The inclination to use `Assert` for any fatal error is understandable, but as I said in my book they are for catching logic errors - not for validating user environment. My non-Bitcoin experience tells me that policies around what should be asserts require upkeep, so I took the time to provide an alternate version.

Was reading https://bitcoincore.academy/bin/onboarding-to-bitcoin-core.html today and came across
> We take extra care during the encryption phase to either complete atomically or fail. This includes database writes where we don’t want to write half and crash, for example. *Therefore we will throw an assertion if the write fails.*

Not sure if the actual code really uses assertions to stop the process if the database reports write failure. Maybe Bitcoin Core has a looser policy than I'm used to, and at least @maflcko is okay with using it in tests.

From a purely functional angle, the `std::cerr` output provides more immediate and actionable information than the assert.
``test/util/random.cpp:29 operator(): Assertion `uint256::FromHex(num)' failed.``
vs
``RANDOM_CTX_SEED must be a 64 char hex string without '0x'-prefix, was set to: '123'.``

I'm happy you incorporated some of my other suggestions but I pointed out this as my major gripe with the PR.

Staying true to my principles and giving this an Approach A-C-K generates cognitive dissonance. But maybe I could be bribed by promises of likely support for a future PR from me cleaning up incorrect `Assert` usage in tests in general.

hodlinator replied:
Thanks for calling out my assumption that `test_runner.py --randomseed=X` could trickle down into unit tests.

Makes my case for it not being a logic error slightly stronger.

stickies-v replied:
Force pushed to use `std::cerr` since it seems everyone can get behind that, thanks for the suggestion.

stickies-v commented:
Thanks a lot for the quick approach feedback, @maflcko. Force pushed to minimize breaking changes by repurposing `IsHexNumber()` to `TrySanitizeHexNumber()` and allowing < 64 characters for `-minimumchainwork` and `RANDOM_CTX_SEED`, and to address the broken `-no<parameter>`  behaviour.

paplorinc commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -32,13 +32,20 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
     }
 
-    if (auto value{args.GetArg("-assumevalid")}) opts.assumed_valid_block = uint256S(*value);
+    if (auto value{args.GetArg("-assumevalid")}; value && *value != "0") {
```
shouldn't we still set the `assumed_valid_block` value when it's `0`, given that `defaultAssumeValid` isn't `0`?

stickies-v replied:
Ugh, you're right, thanks a lot for catching this. This approach would've silently ignored `-noassumevalid` and `-assumevalid=0` by using the `defaultAssumeValid` value instead. I think it's not great that this bug didn't cause any tests to fail. `feature_assumevalid.py` doesn't catch this because `defaultAssumeValid` on regtest [is null](https://github.com/bitcoin/bitcoin/blob/43740f4971f45cd5499470b6a085b3ecd8b96d28/src/kernel/chainparams.cpp#L452) (and I don't think there's a way around this), and we don't seem to have any unittesting on applying args to options.

I've added e9dd47c66426c04fc6f0780d67ba0c05f26f9920 as a first commit (fails on b38a259d1e2749876d4e3a15f218b4f257049320) to add unittesting on `-assumevalid` and `-minimumchainwork`.

paplorinc commented on src/util/strencodings.h:
```diff
@@ -71,9 +71,20 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
+ *        result_size.
+ *
+ * @param input Hex encoding of a number, optionally prefixed with 0x.
+ * @param result_size Size of the resulting string, causing zero-padding
```
I think we could document how to disable this param - currently with 0, I recommend considering negative numbers

l0rinc reviewed (commented):
Thanks for clearing these up, please see my observations.

paplorinc commented on src/test/util_tests.cpp:
```diff
@@ -432,29 +432,33 @@ BOOST_AUTO_TEST_CASE(util_IsHex)
+    BOOST_CHECK(!TrySanitizeHexNumber("0x0x00")); // two prefixes not allowed
+
+    BOOST_CHECK_EQUAL(TrySanitizeHexNumber("0x1234", /*target_size=*/0).value(), "1234");
+    BOOST_CHECK_EQUAL(TrySanitizeHexNumber("0x1234", /*target_size=*/4).value(), "1234");
```
```suggestion
    BOOST_CHECK_EQUAL(TrySanitizeHexNumber("0x1234", /*result_size=*/4).value(), "1234");
```

stickies-v committed:
test: unittest chainstatemanager_args

stickies-v commented on src/util/strencodings.cpp:
```diff
@@ -46,14 +46,19 @@ bool IsHex(std::string_view str)
 }
 
-bool IsHexNumber(std::string_view str)
+std::optional<std::string> TrySanitizeHexNumber(std::string_view input, size_t result_size)
```
The `Number` part indicates that we're dealing with numerical values, where `0x00` means the same thing as `0x0000`, so I think that aligns with how the function behaves?
Return types are not immediately visible on in the callsite, and I think `Try` indicating that the function accepts malformed input is helpful for the reader, so I'm going to leave this as is.

paplorinc replied:
ok, the number part can make sense, I'm still not sure about the `Try` part - seems like https://en.wikipedia.org/wiki/Hungarian_notation, which (as a Hungarian) I'm very much against :p

> Return types are not immediately visible on in the callsite

It's part of the method's signature, I don't see how repeating it in the method's name helps.
The parameters also aren't visible on the call site, yet we're not encoding them in the method's name.

If you insist, I won't block you on this of course, but want to make sure that my arguments against it are clear.

stickies-v commented on src/test/util/random.cpp:
```diff
@@ -7,26 +7,29 @@
-        const char* num = std::getenv(RANDOM_CTX_SEED.c_str());
-        if (num) return uint256S(num);
+        if (const auto num{std::getenv(RANDOM_CTX_SEED)}) {
+            auto sanitized{*Assert(TrySanitizeHexNumber(num, uint256::size() * 2))}; // RANDOM_CTX_SEED must be up to 64 char hex string
```
I couldn't see `RANDOM_CTX_SEED` be documented anywhere as having to be a 64 char hex string, so I thought trying to be lenient would be good for backwards compatibility, but I don't have a view on how important that is so happy to go with your simplification. Adopted in latest force push.

stickies-v commented:
Force-pushed to address review comments, mainly:
- [fixed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1702863593) buggy `-noassumedvalid` behaviour and added unit tests to cover this
- [changed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1702863957) `result_size` to an int type and changed the disabled (default) state to `-1` instead of `0`
- [changed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1703942199) `RANDOM_CTX_SEED` to now have to be a 64 char hex string

stickies-v reviewed (commented):
Force-pushed to address review comments, mainly:
- [fixed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1702863593) buggy `-noassumedvalid` behaviour and added unit tests to cover this
- [changed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1702863957) `result_size` to an int type and changed the disabled (default) state to `-1` instead of `0`
- [changed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1703942199) `RANDOM_CTX_SEED` to now have to be a 64 char hex string

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -46,14 +46,19 @@ bool IsHex(std::string_view str)
-    // Return false for empty string or "0x".
-    return str.size() > 0;
+    std::string result{input};
+    if (input.size() < result_size) {
```
👍, thanks for considering

paplorinc commented:
The CI reminds me of the recent Bitcoin price changes :p

stickies-v commented on src/test/fuzz/hex.cpp:
```diff
@@ -27,7 +27,7 @@ FUZZ_TARGET(hex)
         assert(ToLower(random_hex_string) == hex_data);
     }
-    (void)IsHexNumber(random_hex_string);
+    (void)TrySanitizeHexNumber(random_hex_string);
```
Good idea, updated `fuzz/hex.cpp` to cover `result_size` with an `int16_t`. Not sure what the optimal range of values would be here, but I figured not making it too large (`int`) makes sense given that `result_size` is not user input?

paplorinc commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -769,4 +770,44 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_completion_hash_mismatch, Sna
+
+    // test -assumevalid
+    BOOST_CHECK(!set_opts({}).assumed_valid_block.has_value());
+    BOOST_CHECK(set_opts({"-assumevalid=0"}).assumed_valid_block.value().IsNull());
```
Thanks for covering these scenarios!

I checked to see how this fails for the previous error:
> unknown location:0: fatal error: in "validation_chainstatemanager_tests/chainstatemanager_args": std::bad_optional_access: bad_optional_access

i.e. it doesn't get to the `IsNull` check since it can't call `value()` on a null optional, right?

I think we can work around that by comparing the dereferenced optional instead:
```suggestion
    BOOST_CHECK_EQUAL(*set_opts({"-assumevalid=0"}).assumed_valid_block, uint256::ZERO);
```

which gives this error for the previously invalid setup:
> src/test/validation_chainstatemanager_tests.cpp:801: error: in "validation_chainstatemanager_tests/chainstatemanager_args": check *set_opts({"-assumevalid=0"}).assumed_valid_block == uint256::ZERO has failed [000000016bcc2e20000000016bcc2e20000000016bcc29a03b66800104140f00 != 0000000000000000000000000000000000000000000000000000000000000000]

Which I find a lot more revealing.

paplorinc replied:
> Can you run this in valgrind?

Thanks for the context @maflcko! Valgrind is currently isn't really [compatible with ARM-based Macs](https://github.com/LouisBrunner/valgrind-macos/issues/56).

> Dereferencing a nullopt is UB 

And we can't use `BOOST_CHECK_EQUAL(set_opts({"-assumevalid=0"}).assumed_valid_block, std::make_optional(uint256::ZERO))` since there's no standard way to display optionals, right?

Ok, in that case just consider the test fix validated, it failed before the fix, passes now.

paplorinc replied:
I've noted it and will give it a try next time, thank you

paplorinc replied:
Thanks, I think that would produce better error messages on failure.

stickies-v replied:
Would be nice to have std::optional support in BOOST_CHECK_EQUAL but this PR has already had quite a bit of churn so I'm going to limit the scope and leave as is given that it's not super relevant.

paplorinc replied:
Split out to pr [`aec1c90` (#30618)](https://github.com/bitcoin/bitcoin/pull/30618/commits/aec1c901118311eec5dc24104a7439377c8fd6de)

paplorinc commented on src/test/util_tests.cpp:
```diff
@@ -432,29 +432,34 @@ BOOST_AUTO_TEST_CASE(util_IsHex)
+    BOOST_CHECK_EQUAL(TrySanitizeHexNumber("0x1234", /*result_size=*/-1).value(), "1234");
+    BOOST_CHECK_EQUAL(TrySanitizeHexNumber("0x1234", /*result_size=*/4).value(), "1234");
+    BOOST_CHECK_EQUAL(TrySanitizeHexNumber("1234", /*result_size=*/5).value(), "01234");
+    BOOST_CHECK(!TrySanitizeHexNumber("0x1234", /*result_size=*/0));
```
nice

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -46,14 +46,20 @@ bool IsHex(std::string_view str)
-    for (char c : str) {
-        if (HexDigit(c) < 0) return false;
+    input = util::RemovePrefixView(input, "0x");
+    if (result_size < 0) result_size = input.size(); // negative result_size disables the size check
```
nit: my first reaction was here - "I wonder if there's an off-by-one error here".
If you think it makes sense, consider this alternative:
```suggestion
    if (result_size < 0) result_size = std::numeric_limits<int>::max(); // negative result_size disables the size check
```

stickies-v replied:
That wouldn't work. `result_size` is used both to throw if the input is too large, as well as to pad with leading zeroes if it's too short: https://github.com/bitcoin/bitcoin/pull/30569/files#diff-3f688af8f182edecd9c33977b905b3e71dc010574f721fd4e328bdbc7706f574R59

paplorinc replied:
K, in that case the test cases check the boundaries correctly already, thanks for considering.

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -46,14 +46,20 @@ bool IsHex(std::string_view str)
-        if (HexDigit(c) < 0) return false;
+    input = util::RemovePrefixView(input, "0x");
+    if (result_size < 0) result_size = input.size(); // negative result_size disables the size check
+    if (input.empty() || ((int)input.size() > result_size)) return std::nullopt;
```
I don't particularly  like the alternative, but the developer notes state:
```
- Use a named cast or functional cast, not a C-Style cast. When casting
    between integer types, use functional casts such as `int(x)` or `int{x}`
    instead of `(int) x`. When casting between more complex types, use `static_cast`.
 ```

stickies-v replied:
Thanks, will change to `int{x}` in next force push.

paplorinc replied:
That might not work, but extracting might:
```C++
    auto final_size = (result_size < 0) ? input.size() : static_cast<size_t>(result_size);
```

l0rinc reviewed (commented):
Will continue reviewing a bit later, I only had time for these so far

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -46,14 +46,20 @@ bool IsHex(std::string_view str)
+    input = util::RemovePrefixView(input, "0x");
+    if (result_size < 0) result_size = input.size(); // negative result_size disables the size check
+    if (input.empty() || ((int)input.size() > result_size)) return std::nullopt;
+    for (char c : input) {
+        if (HexDigit(c) < 0) return std::nullopt;
     }
```
could we use `IsHex` here instead, i.e. merge it with previous line:
```C++
    if (input.empty() || ((int)input.size() > result_size) || !IsHex(input)) return std::nullopt;
```

though this contains an extra evennes requirement which we may not want here.

paplorinc commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -32,10 +32,11 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
 
     if (auto value{args.GetArg("-minimumchainwork")}) {
-        if (!TrySanitizeHexNumber(*value)) {
+        if (auto sanitized_hex{TrySanitizeHexNumber(*value, /*result_size=*/64)}) {
```
in the next commit we're using `uint256::size() * 2` instead of the hard-coded 64 - consider unifying, if you agree that it documents the behavior slightly better

paplorinc commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -39,7 +39,14 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
 
-    if (auto value{args.GetArg("-assumevalid")}) opts.assumed_valid_block = uint256S(*value);
+    if (auto value{args.GetArg("-assumevalid")}) {
+        if (*value == "0") value.emplace(uint256::size() * 2, '0'); // handle -noassumevalid
```
nit: if you think it's more readable, here's an alternative which avoids reparsing and duplicating `uint256::size() * 2` (but assigns twice, so you may not like it):
```C++
    if (auto value{args.GetArg("-assumevalid")}) {
        if (*value == "0") { // handle -noassumevalid
            opts.assumed_valid_block = uint256{};
        } else if (auto block_hash{uint256::FromHex(*value)}) {
            opts.assumed_valid_block = *block_hash;
        } else {
            return util::Error{strprintf(Untranslated("Invalid assumevalid block hash specified (%s), must be %d character hex (or 0 to disable)"), *value, uint256::size() * 2)};
        }
    }
```

l0rinc reviewed (commented):
Thanks, please see my remaining suggestions

stickies-v reviewed (commented):
Thanks for the review @paplorinc, force pushed to address all outstanding comments. Style-only changes:
- [introduced](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1706884207) `final_size` var to avoid typecasting between (un)signed types and some related cleanup
- [removed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1706891172) hardcoded `64` value
- [cleaned up](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1706897088) the `-noassumevalid` logic a bit

l0rinc reviewed (approved):
ACK d045fc7ac1729cf29140c43518f20375f2aaa1cc

Note that I don't know the second order effects of constraining previous public input values, somebody else should assess that part

hodlinator commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -769,4 +770,44 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_completion_hash_mismatch, Sna
+    argv.insert(argv.end(), args.begin(), args.end());
+
+    std::string error{};
+    BOOST_REQUIRE(args_man.ParseParameters(argv.size(), argv.data(), error));
```
nit:
```suggestion
    BOOST_REQUIRE(args_man.ParseParameters(argv.size(), argv.data(), error));
    BOOST_REQUIRE(error.empty());
```

stickies-v replied:
This feels like a distraction to me tbh, we're not unit testing `ParseParameters` here. I think it's safe to rely on `ParseParameters` returning `false` when there's an error.

hodlinator reviewed (changes requested):
Concept ACK d045fc7ac1729cf29140c43518f20375f2aaa1cc

Thanks for chipping away at making hex string handling more robust!

Some minor disagreements + ignorable nits. The most significant disagreement in this thread: https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1700718302

Tested:
```bash
$ RANDOM_CTX_SEED=1231231231231231231231231231231231231231231231231231231231312312 src/test/test_bitcoin
$ test/functional/feature_assumevalid.py
```

maflcko reviewed (approved):
left a nit. Feel free to ignore.

Otherwise this looks good on a first glance.

stickies-v commented on src/test/util/random.cpp:
```diff
@@ -7,26 +7,27 @@
+        }
         // Otherwise use a (truly) random value.
         return GetRandHash();
     }();
```
Done, I've taken your commit (hadn't seen the commit message earlier, very nice) but just added the docstring (and smaller `num` scope) back in.

stickies-v commented on src/test/util/random.cpp:
```diff
@@ -7,26 +7,27 @@
 
 FastRandomContext g_insecure_rand_ctx;
 
+inline constexpr const char* RANDOM_CTX_SEED{"RANDOM_CTX_SEED"};
```
I think you're right that `inline` was not necessary here. No longer applies as I've adopted [maflcko's suggestion](https://github.com/bitcoin/bitcoin/pull/30569/files#r1707958805) which moves the var back inside the `SeedRandomForTest` scope.

stickies-v commented on src/util/strencodings.cpp:
```diff
@@ -46,14 +46,19 @@ bool IsHex(std::string_view str)
+    std::string result{input};
+    if (auto padding{final_size - input.size()}; padding > 0) {
+        result = std::string(padding, '0') + result;
+    }
```
Don't think this is worth spending too much time on but given that the only non-test call is quite likely to require padding, I've updated it to:

```diff
diff --git a/src/util/strencodings.cpp b/src/util/strencodings.cpp
index 4e0317cd0e..be946af269 100644
--- a/src/util/strencodings.cpp
+++ b/src/util/strencodings.cpp
@@ -8,6 +8,7 @@
 #include <crypto/hex_base.h>
 #include <span.h>
 
+#include <algorithm>
 #include <array>
 #include <cassert>
 #include <cstring>
@@ -54,10 +55,8 @@ std::optional<std::string> TrySanitizeHexNumber(std::string_view input, int resu
     for (char c : input) {
         if (HexDigit(c) < 0) return std::nullopt;
     }
-    std::string result{input};
-    if (auto padding{final_size - input.size()}; padding > 0) {
-        result = std::string(padding, '0') + result;
-    }
+    std::string result(final_size, '0');
+    std::copy(input.begin(), input.end(), result.begin() + (final_size - input.size()));
     return result;
 }
 


```

stickies-v commented on src/util/strencodings.h:
```diff
@@ -71,9 +71,20 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
+ *                    if input is too short, or a std::nullopt return
+ *                    value if input is too long. Disabled if <= -1.
+ *                    (default: -1)
+ * @return std::optional<std::string>
```
Agreed, autogenerated and didn't think to remove it. Done.

stickies-v commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -32,7 +32,7 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
 
     if (auto value{args.GetArg("-minimumchainwork")}) {
-        if (!IsHexNumber(*value)) {
+        if (!TrySanitizeHexNumber(*value)) {
```
I disagree, a227cb511ec948b37ddbc9ee65de586109ebc1da is a refactor commit so I prefer keeping the behaviour-changing commits such as e6f81b9d81359a7ffeff6d1830188f00df0e8db0 smaller and separate.

stickies-v commented:
Force-pushed to address all outstanding review comments, mainly:
- [avoid string re-allocation](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1707728879)
- [cleaned up](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1707958805) `util/random.cpp` a bit more
- [removed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1707734248) unnecessary docstring

stickies-v reviewed (commented):
Force-pushed to address all outstanding review comments, mainly:
- [avoid string re-allocation](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1707728879)
- [cleaned up](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1707958805) `util/random.cpp` a bit more
- [removed](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1707734248) unnecessary docstring

paplorinc commented on src/util/strencodings.cpp:
```diff
@@ -46,14 +47,17 @@ bool IsHex(std::string_view str)
-    if (str.substr(0, 2) == "0x") str.remove_prefix(2);
-    for (char c : str) {
-        if (HexDigit(c) < 0) return false;
+    input = util::RemovePrefixView(input, "0x");
+    const auto final_size{(result_size < 0) ? input.size() : static_cast<size_t>(result_size)};
+    if (input.empty() || (input.size() > final_size)) return std::nullopt;
+    for (char c : input) {
+        if (HexDigit(c) < 0) return std::nullopt;
     }
-    // Return false for empty string or "0x".
-    return str.size() > 0;
+    std::string result(final_size, '0');
+    std::copy(input.begin(), input.end(), result.begin() + (final_size - input.size()));
+    return result;
 }
```
Given that:
* this is only used in a single place in prod (which accept `std::string_view` as well and has an `std::string` input)
* we're always copying the result before returning
* the overwhelming majority of inputs will probably have the correct size already, i.e. don't need change.
* the performance of the method seems to be important

Could we either return an `std::optional<std::string_view>` instead or take an `std::string& input` to be able to return it?

i.e.
```suggestion
std::optional<std::string> TrySanitizeHexNumber(std::string input, int result_size)
{
    input = util::RemovePrefixView(input, "0x");
    const auto final_size{(result_size < 0) ? input.size() : static_cast<size_t>(result_size)};
    if (input.empty() || (input.size() > final_size)) return std::nullopt;
    for (char c : input) {
        if (HexDigit(c) < 0) [[unlikely]] return std::nullopt;
    }
    if (input.size() == final_size) [[likely]] return input;

    std::string result(final_size, '0');
    std::copy(input.begin(), input.end(), result.begin() + (final_size - input.size()));
    return result;
}
```

hodlinator replied:
I agree taking `input` as `std::string` works better with the return type.

Instead of introducing a 2nd string at the end, one could simply:
```C++
    input.insert(0, final_size - input.size(), '0');
    return input;
```

paplorinc replied:
>    input.insert(0, final_size - input.size(), '0');
    return input;

my functional programming past screams at the though of mutating parameters and also returning them
<insert meme with PTSD Chihuahua>

hodlinator replied:
Yeah, I had a bit of an itch as I was writing `return input`, but you made your bed when you started taking `std::string` by value. :)

stickies-v replied:
> the performance of the method seems to be important

How so? There is one callsite, which can be called at most once, to parse one optional debug-only option.

> Could we either return an std::optional<std::string_view>

How would that work? `std::string_view` is a non-owning view.

> we're always copying the result before returning

In 802374b4355bd1dec7a88bba6287c55f935699fe, we're always allocating exactly once. In your example, we'd be allocating at least once, and twice if we need padding. That doesn't look like an improvement to me.

> the overwhelming majority of inputs will probably have the correct size already, i.e. don't need change.

Given that it relies on user input, that seems hard to substantiate. In the functional tests, for example, we're using quite a bit of non-64 char hex input, which is reasonable given that we're doing all this work to accept number input.

hodlinator replied:
```C++
input = util::RemovePrefixView(input, "0x");
```
This line gives me the heebie-jeebies if input is `std::string`. It might be okay as we are assigning a shorter string, but worst case, the assignment operator could free it's prior heap buffer and allocate a shorter buffer, before attempting to copy from the string_view which now might reference the deallocated buffer. Using `RemovePrefix()` should alleviate the problem.

paplorinc replied:
Will let you decide, there are multiple ways to do this, each would be slightly better in a different scenario.
Whichever produces the cleanest code, choose that :)

stickies-v commented:
Force pushed to [terminate with `std::abort()`](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1709568559) in case of invalid `RANDOM_CTX_SEED` input.

stickies-v commented:
> I think the description can be clarified that there are two user-facing settings that are now stricter checked.

Thanks, added a section on introduced behaviour change.

hodlinator commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -32,10 +32,11 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
     if (auto value{args.GetArg("-minimumchainwork")}) {
-        if (!TrySanitizeHexNumber(*value)) {
+        if (auto sanitized_hex{TrySanitizeHexNumber(*value, /*result_size=*/uint256::size() * 2)}) {
+            opts.minimum_chain_work = UintToArith256(*uint256::FromHex(*sanitized_hex));
```
nit: An `Assume()` or `Assert()` around the `FromHex()` result before the `*`-deref might be better here, even though it should work :tm:.

Maybe linking `TrySanitizeHexNumber` and `FromHex` together in the fuzz-test would be good for enforcing that the former only allows through something acceptable to the latter (although `result_size` must also be correct).

stickies-v replied:
Yeah, even though this assumption seems safe at the moment it's brittle and blindly dereferencing it is not good practice. Will use `.value()` either in next force-push or in the follow-up PR removing `uint256S` completely. Thanks.

paplorinc replied:
> blindly dereferencing it is not good practice

In C++23 we'll be able to chain them properly via something like `auto hex_value = sanitized_hex.and_then(uint256::FromHex);`
If there's a C++23 checklist somewhere (haven't seen any), we could just leave the dereference as is and fix it after we migrate to C++23 instead.

stickies-v replied:
> (In the follow-up you can also adjust the error message to explain the new max-length check.

Done.

> Will use .value() either in next force-push

Resolved by using `FromUserHex()` instead.

hodlinator reviewed (approved):
ACK 855784d3a0026414159acc42fceeb271f8a28133

Passed `make check`.
Passed ``RANDOM_CTX_SEED=1231231231231231231231231231231231231231231231231231231231312312 src/test/test_bitcoin``.
Failed with expected error when removing char or inserting non-hex char into `RANDOM_CTX_SEED` from above.

Thanks @stickies-v for humoring me regarding the error handling in **test/util/random.cpp**!

paplorinc commented on src/test/fuzz/hex.cpp:
```diff
@@ -19,15 +20,17 @@
         assert(ToLower(random_hex_string) == hex_data);
     }
-    (void)IsHexNumber(random_hex_string);
+    (void)TrySanitizeHexNumber(random_hex_string, result_size);
```
Split out to other pr: [`fd6d110` (#30618)](https://github.com/bitcoin/bitcoin/pull/30618/commits/fd6d1106c07d009a9f9517d905feee6f4136d8d7)

achow101 commented:
> `-assumevalid` will raise an error when input contains invalid hex characters (including `0x` prefix), or when it is not exactly 64 characters long

Why not allow `0x` prefixes as `-minimumchainwork` does? It seems plausible to me that someone might have put a `0x` prefixed block hash in their conf file and this change breaks that.

ryanofsky reviewed (commented):
Code review 855784d3a0026414159acc42fceeb271f8a28133, mild NACK.

I like the new test, and ideas behind this change, but I think some parts of this are not well documented, and are breaking things that work without a clear reason.

stickies-v commented on src/util/strencodings.h:
```diff
@@ -71,9 +71,19 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
+ *        result_size.
+ *
+ * @param input Hex encoding of a number, optionally prefixed with 0x.
+ * @param result_size Size of the resulting string, causing zero-padding
```
> Would be good to change "Size of" to "Desired size of".

"Desired" to me sounds like the result string may not actually be of the requested size, when it is actually guaranteed. I'm not sure that would be an improvement?

stickies-v replied:
Resolved by removing `TrySanitizeHexNumber()` in latest force-push.

stickies-v commented on src/util/strencodings.h:
```diff
@@ -71,9 +71,19 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
-* Return true if the string is a hex number, optionally prefixed with "0x"
-*/
-bool IsHexNumber(std::string_view str);
+ * @brief Return a sanitized hex string if input is a valid hex number.
+ *        "0x" prefix is removed, and if result_size is specified, the
+ *        result is padded with leading zeroes until result_size is
+ *        reached, or std::nullopt returned if input is longer than
+ *        result_size.
```
> I think instead of saying this "sanitizes" the string and pretending like that word means something it would be better to just say this strips 0x prefixes and fixes the length of the string so it can be passed to the base_blob::ParseHex() function.

Yeah, I felt a bit uneasy about the arbitrariness of what the function is doing and how it's named. I can't really come up with a better name, though, but I'm open to suggestions?

An alternative would be to, as @maflcko initially [suggested](https://github.com/bitcoin/bitcoin/pull/30569/#discussion_r1700712603), just have this be a parameter of `base_blob::FromHex()`, but that feels less elegant to me, still. What do you think?

stickies-v replied:
> It's hard to imagine this function being used for anything other than as a preprocessor to `FromHex`

The reason I carved it out instead of having it be a `FromHex()` parameter is not for reusability but to not bloat `FromHex()` with an explosion of input validation options (which also leaks into functions wrapping `base_blob::FromHex()`, such as `transaction_identifier::FromHex()`). I don't think there is a single answer to what we'll ever need for e.g. the below options, it can very much vary for each callsite and context:
- do we accept leading and/or trailing whitespaces
- do we require fixed size input
- do we forbid, allow or require a 0x prefix
- is an empty string forbidden or equivalent to 0

I think having an opinionated input validation function (and potentially more in the future, when the need arises) and a pristine `FromHex()` that does not concern itself with input validation is a much more maintainable and testable approach. For that reason, I'd prefer not parameterizing `FromHex()`.

Would an approach like the below (moved into the `uint256.h` header), alleviate your concern?

```
/**
 * @brief Helper function for \ref base_blob::FromHex. Returns a copy of
 *        the input without (optional) "0x" prefix which is padded with
 *        leading zeroes if it is shorter than what T::FromHex expects.
 *        If input contains invalid hex characters, including
 *        whitespace, or if it is longer than what T::FromHex expects, a
 *        std::nullopt is returned.
 *
 *        If a value is returned, it will be accepted by T::FromHex().
 *
 * @param input Hex encoding of a number, optionally prefixed with 0x.
 */
template <typename T>
std::optional<std::string> ValidateAndPadHexNumber(std::string_view input) 
    requires std::is_base_of_v<base_blob<T::size() * 8>, T>;
```

ryanofsky replied:
I think I would probably add `base_blob::FromUserHex()` function wrapping `base_blob::FromHex()`. `FromHex()` would require strict formatting with fixed width strings and no extraneous characters. But when provided with user input, `FromUserHex()` would allow `0x` prefixes, and not require numbers to be padded to a fixed number of digits. Either way, numbers that are too big to fit in the blob or contained unrecognized characters would be errors.

There is no need for either of these functions to trim whitespace as we already have convenient functions to do that, and decision on whether to allow whitespace depends more on how the string arrived (from a file, or over the command line or RPC) than on what produced it.

I think a `FromUserHex` function would be better than a sidecar `ValidateAndPadHexNumber` function because it would be easier to use, and more discoverable. The more it was used, the more it would provide a consistent user interface that did not arbitrarily reject things some places which are accepted other places.

stickies-v replied:
> I think I would probably add base_blob::FromUserHex() function wrapping base_blob::FromHex().

Thanks for the suggestion, I've adopted this approach.

stickies-v reviewed (commented):
> Why not allow 0x prefixes as -minimumchainwork does? 

Because my understanding is that `0x` is used for hex numbers, and a block hash is not a number. So we have to make a trade-off between not introducing "weirdness" into the code vs maintaining backwards compatibility (for an undocumented feature). I prefer leaning towards the former unless there are good reasons to go for the latter, but that's exactly the reason I carved out these commits into a smaller PR so thank you for your input.

paplorinc commented:
> Because my understanding is that `0x` is used for hex numbers, and a block hash is not a number.

If it's not a number, why are we prefixing it with 0 (by a method called TrySanitizeHex*Number*)?

> parameter of base_blob::FromHex(),

or alternatively chain the two together like `TrimAndParse`, i.e. `base_blob::FromHexLenient()`

> -assumedvalid [...] -minimumchainwork

Would it be clearer if we just warned about the changes in this release and require sanitized input in the next?

ryanofsky commented on src/test/util/random.cpp:
```diff
@@ -7,32 +7,39 @@
-        const char* num = std::getenv(RANDOM_CTX_SEED.c_str());
-        if (num) return uint256S(num);
+        if (const char* num{std::getenv(RANDOM_CTX_SEED)}) {
+            if (auto num_parsed{uint256::FromHex(num)}) {
```
re: https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1712563888

Thanks for the links. I don't think I realized this change was made in the other PR, and I think it is not a good change.

I don't see a reason somebody shouldn't be able to run `RANDOM_CTX_SEED=12345 ./test_bitcoin` conveniently from a command line or script without padding the seed with 59 0's. It seems like this makes it pointlessly difficult to specify a random seed and I don't get the motivation behind it.

If we want to go ahead with change, I think it should be described accurately as making something that used to work no longer work, not just as adding error checking and implying that anything this now rejects was previously in error.

stickies-v replied:
> It seems like this makes it pointlessly difficult to specify a random seed and I don't get the motivation behind it.

I definitely didn't mean to introduce user friction with this PR - the philosophy was more to not introduce unnecessary code complexity for features that aren't used, but since it seems like this might be breaking existing/valuable use cases, I've now updated `RANDOM_CTX_SEED` to use the more lenient `uint256::FromUserHex()` to still allow 0x-prefixes and shorter input, and it doesn't add code complexity anyway.

ryanofsky reviewed (commented):
Code review 855784d3a0026414159acc42fceeb271f8a28133 again (unchanged since last time). I think motivation here is good and this PR could be merged if my concerns are not shared, but IMO this change is a net negative in current state without some compatibility improvements.

re: https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1712563888

> Because my understanding is that `0x` is used for hex numbers, and a block hash is not a number.

I do think block hashes are numbers, and it's not really unusual to treat them that way, but even if it is strange and bad to think of them as numbers, I would want to give more thought to backwards compatibility, and only make things errors when they are (1) definitely errors or (2) have a pretty likelihood of being errors, and not just shorter or different styles of writing things.

hodlinator commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -39,7 +39,15 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
+    if (auto value{args.GetArg("-assumevalid")}) {
+        if (*value == "0") { // handle -noassumevalid
+            opts.assumed_valid_block = uint256{};
+        } else if (auto block_hash{uint256::FromHex(*value)}) {
```
I didn't previously realize that this PR breaks use-facing compatibility wrt `0x` prefixes in args. Would also like to see such cases be filtered through `TrySanitizeHexNumber` before going into `FromHex` so that level of compatibility is maintained.

stickies-v replied:
Thank you for summarizing your concerns @ryanofsky. I believe the latest force-push addresses all of them.

> It's good this change is making it an error to include trailing non-hex characters instead of silently ignoring them.
It's good this change is making it an error to specify hex strings that are too long.
It's probably ok this change makes it an error to include leading and trailing whitespace, instead of trimming it. Maybe that could be safer and prevent configuration bugs.

Relevant behaviour unchanged in latest force-push.

> It's not good but probably ok this change makes it an error to specify an empty string (-assumedvalid="") to disable the assumevalid feature. That seemed like a useful way to clear the setting.
It's not good this change makes it an error to include a 0x prefix or write numbers not padded with 0's. That just seems like being arbitrarily strict and breaking backwards compatibility for no reason.

Latest force-push continues to allow `-assumevalid=""` as well as 0x-prefixed and too-short-input (and unit test cases are added to `test: unittest chainstatemanager_args`), keeping this behaviour the same as it currently is on master.

hodlinator reviewed (changes requested):
Approach NACK 855784d3a0026414159acc42fceeb271f8a28133

..in light of https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1714952020

hodlinator commented:
Would be great to get the final issues resolved. Possible way forward, preserving backwards compatibility:

<details>
<summary>
diff
</summary>

```diff
diff --git a/src/node/chainstatemanager_args.cpp b/src/node/chainstatemanager_args.cpp
index ff5a7ebd00..e83b5f38b6 100644
--- a/src/node/chainstatemanager_args.cpp
+++ b/src/node/chainstatemanager_args.cpp
@@ -33,7 +33,7 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
 
     if (auto value{args.GetArg("-minimumchainwork")}) {
         if (auto sanitized_hex{TrySanitizeHexNumber(*value, /*result_size=*/uint256::size() * 2)}) {
-            opts.minimum_chain_work = UintToArith256(*uint256::FromHex(*sanitized_hex));
+            opts.minimum_chain_work = UintToArith256(uint256::FromHex(*sanitized_hex).value());
         } else {
             return util::Error{strprintf(Untranslated("Invalid non-hex (%s) minimum chain work value specified"), *value)};
         }
@@ -42,10 +42,10 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
     if (auto value{args.GetArg("-assumevalid")}) {
         if (*value == "0") { // handle -noassumevalid
             opts.assumed_valid_block = uint256{};
-        } else if (auto block_hash{uint256::FromHex(*value)}) {
-            opts.assumed_valid_block = *block_hash;
+        } else if (auto sanitized_hex{TrySanitizeHexNumber(*value, /*result_size=*/uint256::size() * 2)}) {
+            opts.assumed_valid_block = uint256::FromHex(*sanitized_hex).value();
         } else {
-            return util::Error{strprintf(Untranslated("Invalid assumevalid block hash specified (%s), must be %d character hex (or 0 to disable)"), *value, uint256::size() * 2)};
+            return util::Error{strprintf(Untranslated("Invalid assumevalid block hash specified (%s), must be %d hex digits (or 0 to disable)"), *value, uint256::size() * 2)};
         }
     }
 
diff --git a/src/test/util/random.cpp b/src/test/util/random.cpp
index 0c09d8fd24..26f114c6d4 100644
--- a/src/test/util/random.cpp
+++ b/src/test/util/random.cpp
@@ -27,8 +27,8 @@ void SeedRandomForTest(SeedRand seedtype)
     static const uint256 ctx_seed = []() {
         // If RANDOM_CTX_SEED is set, use that as seed.
         if (const char* num{std::getenv(RANDOM_CTX_SEED)}) {
-            if (auto num_parsed{uint256::FromHex(num)}) {
-                return *num_parsed;
+            if (auto num_sanitized{TrySanitizeHexNumber(num, /*result_size=*/uint256::size() * 2)}) {
+                return uint256::FromHex(*num_sanitized).value();
             } else {
                 std::cerr << RANDOM_CTX_SEED << " must be a " << uint256::size() * 2 << " char hex string without '0x'-prefix, was set to: '" << num << "'.\n";
                 std::abort();
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index 6f81444aa9..63b687a3c4 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -802,6 +802,8 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
     BOOST_CHECK(set_opts({"-noassumevalid"}).assumed_valid_block.value().IsNull());
     const std::string cmd{"-assumevalid=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"};
     BOOST_CHECK_EQUAL(set_opts({cmd.c_str()}).assumed_valid_block.value().ToString(), cmd.substr(13, cmd.size()));
+    const std::string prefixed{"-assumevalid=0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"};
+    BOOST_CHECK_EQUAL(set_opts({prefixed.c_str()}).assumed_valid_block.value().ToString(), prefixed.substr(15, prefixed.size()));
 
     // test -minimumchainwork
     BOOST_CHECK(!set_opts({}).minimum_chain_work.has_value());
```
</details>

Side note: `make check` fails for me in 2 steps after the regular unit tests, same thing happens on the base commit of the PR, so it's not caused by the PR changes.

stickies-v committed:
refactor: add uint256::FromUserHex helper

FromUserHex will be used in future commits to construct
uint256 instances from user hex input without being
unnecessarily restrictive on formatting by allowing
0x-prefixed input that is shorter than 64 characters.

stickies-v committed:
util: remove unused IsHexNumber

The relevant unit tests have been incorporated in
uint256_tests/from_user_hex in a previous commit.

l0rinc commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -32,13 +32,20 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
+        if (auto min_work{uint256::FromUserHex(*value)}) {
+            opts.minimum_chain_work = UintToArith256(*min_work);
+        } else {
+            return util::Error{strprintf(Untranslated("Invalid minimum chain work value specified (%s), must be up to %d characters hex"), *value, uint256::size() * 2)};
```
Nit (feel free to disregard): Based on the description (`Minimum work assumed to exist on a valid chain in hex`), "chain" and "value" might be superfluous here
```suggestion
            return util::Error{strprintf(Untranslated("Invalid minimum work specified (%s), must be up to %d characters hex"), *value, uint256::size() * 2)};
```

stickies-v replied:
Taken both suggestions in latest force-push, thanks.

l0rinc commented on src/uint256.h:
```diff
@@ -106,6 +108,25 @@ std::optional<uintN_t> FromHex(std::string_view str)
+ *        Designed to be used when dealing with user input.
+ */
+template <class uintN_t>
+std::optional<uintN_t> FromUserHex(std::string_view input)
```
Since this is a "User" centric specialization of `FromHex` (and not a `Hex` specialization of `FromUser`), I'd keep the `FromHex` prefix, i.e.: `FromHexUser` or perhaps `FromHexLenient`, since the `User` part is just incidental

stickies-v replied:
I have nothing against any of the suggested names, but personally I still find `FromUserHex` to be the most intuitive (and it seems 2 at least 2 other reviewers like it as well), so I'm going to stick with that if that's okay?

l0rinc commented on src/uint256.h:
```diff
@@ -106,6 +108,25 @@ std::optional<uintN_t> FromHex(std::string_view str)
+std::optional<uintN_t> FromUserHex(std::string_view input)
+{
+    input = util::RemovePrefixView(input, "0x");
+    constexpr auto expected_size{uintN_t::size() * 2};
```
Could we extract this this as a constant and use it throughout this PR (e.g. in error messages and tests)?

hodlinator replied:
(Thread https://github.com/bitcoin/bitcoin/pull/30569/files#r1725694129)
(This has been silently nagging me as well, but totally understand if you want to leave it for someone to follow up on).

stickies-v replied:
I'm open to the idea, but I don't immediately see a super elegant solution. Thoughts:
- a lot of the `::size() * 2` uses are templated (such as in the referenced example here), so constants wouldn't really work. Adding a `base_blob::hex_size()` seems inappropriate to me too, that's not really a `base_blob` concern.
- non-templated use is not very frequent, and mostly in tests. Potentially a `constexpr auto UINT256_HEX_SIZE{uint256::size() * 2}` could work, but I'm not really convinced that's worth it either tbh

l0rinc replied:
Thanks for considering

l0rinc commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -769,4 +772,59 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_completion_hash_mismatch, Sna
+
+    // test -assumevalid
+    BOOST_CHECK(!get_valid_opts({}).assumed_valid_block.has_value());
+    BOOST_CHECK(get_valid_opts({"-assumevalid="}).assumed_valid_block.value().IsNull());
```
Are we sure this behavior should be written in stone? It rather seems like an error to me

stickies-v replied:
This behaviour was mentioned [here](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1714546457) (and correctly identified as broken by a previous version of this PR), so that's why I added a test to make it explicit. I don't have a strong view on whether that behaviour should be allowed, but since it exists, and some people seem to find it useful, and we're trying to minimize breaking behaviour now, I think I'd prefer keeping it and documenting it.

l0rinc replied:
Haven't realized it was possible before as well, thanks for clearing that up

l0rinc commented on src/test/uint256_tests.cpp:
```diff
@@ -386,6 +386,33 @@ BOOST_AUTO_TEST_CASE(from_hex)
 
+BOOST_AUTO_TEST_CASE(from_user_hex)
+{
+    constexpr std::string_view valid_hex_65{"0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0"}; // 0x prefix and 65 hex chars
```
could we move this closer to first usage?

stickies-v replied:
Sure, done!

l0rinc commented on src/test/uint256_tests.cpp:
```diff
@@ -386,6 +386,33 @@ BOOST_AUTO_TEST_CASE(from_hex)
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("").value(), uint256::ZERO);
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0x").value(), uint256::ZERO);
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0").value(), uint256::ZERO);
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0x10").value(), uint256{16});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("10").value(), uint256{16});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0xff").value(), uint256{255});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("ff").value(), uint256{255});
```
```suggestion
    BOOST_CHECK_EQUAL(uint256::FromUserHex("0x10").value(), uint256{0x10});
    BOOST_CHECK_EQUAL(uint256::FromUserHex("10").value(), uint256{0x10});
    BOOST_CHECK_EQUAL(uint256::FromUserHex("0xff").value(), uint256{0xff});
    BOOST_CHECK_EQUAL(uint256::FromUserHex("ff").value(), uint256{0xff});
```

stickies-v replied:
Thanks, that makes much more sense indeed, taken.

l0rinc commented on src/test/uint256_tests.cpp:
```diff
@@ -386,6 +386,33 @@ BOOST_AUTO_TEST_CASE(from_hex)
+
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("").value(), uint256::ZERO);
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0x").value(), uint256::ZERO);
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0").value(), uint256::ZERO);
```
nit: I was wondering if it accepts `00` as well (since numbers usually trim leading zeros) and if it accepts `1` (since other implementations require an even number of digits), consider extending the examples:
```C++
    BOOST_CHECK_EQUAL(uint256::FromUserHex("00").value(), uint256::ZERO);
    BOOST_CHECK_EQUAL(uint256::FromUserHex("1").value(), uint256::ONE);
```

stickies-v replied:
Thanks, I've added these two test cases!

hodlinator reviewed (approved):
ACK cf88a55e97719aabd62f0b608df0800fef8304de

I really like the `TrySanitizeHexNumber` -> `FromUserHex` change!
Sorry for not understanding that you had it in the works with my prior suggestion.
Only ignorable nits left.

(Still fails at the end of `make check` due to unlucky base commit).

stickies-v commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -769,4 +772,59 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_completion_hash_mismatch, Sna
+    BOOST_CHECK(get_valid_opts({"-assumevalid=0"}).assumed_valid_block.value().IsNull());
+    BOOST_CHECK(get_valid_opts({"-noassumevalid"}).assumed_valid_block.value().IsNull());
+    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0x1234"}).assumed_valid_block.value().ToString(), std::string(60, '0') + "1234");
+    const std::string cmd{"-assumevalid=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"};
+    BOOST_CHECK_EQUAL(get_valid_opts({cmd.c_str()}).assumed_valid_block.value().ToString(), cmd.substr(13, cmd.size()));
```
Yeah, I'm trying to keep tests focused on the specific function under test. We already have proper "0x" prefix testing in `uint256_tests/from_user_hex` tests, and as belt-and-suspenders this case also covers a bit of both but I'd rather not bloat it too much.

stickies-v commented on src/test/util/random.cpp:
```diff
@@ -7,32 +7,39 @@
 
     const uint256& seed{seedtype == SeedRand::SEED ? ctx_seed : uint256::ZERO};
-    LogPrintf("%s: Setting random seed for current tests to %s=%s\n", __func__, RANDOM_CTX_SEED, seed.GetHex());
+    LogPrintf("Setting random seed for current tests to %s=%s\n", RANDOM_CTX_SEED, seed.GetHex());
```
Oops, it wasn't, thanks for spotting. Reverted in latest force-push.

stickies-v commented on src/util/strencodings.h:
```diff
@@ -71,9 +71,19 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
+ *        "0x" prefix is removed, and if result_size is specified, the
+ *        result is padded with leading zeroes until result_size is
+ *        reached, or std::nullopt returned if input is longer than
+ *        result_size.
```
Resolved by removing `TrySanitizeHexNumber()` altogether (and empty strings are still allowed in `uint256::FromUserHex`, as per current behaviour on `master`)

stickies-v reviewed (commented):
Force-pushed (twice) to change the approach to address @achow101's, @ryanofsky's and @hodlinator's concerns about being too strict about user hex input for `-assumevalid` and `RANDOM_CTX_SEED`. Thanks again for sharing your concerns about breaking changes (and the detailed review), in hindsight I think this is indeed the better approach.

Specifically, these last two force-pushes:
- introduces a `uint256::FromUserHex()` to allow for lenient user input parsing and
does away with the `TrySanitizeHexNumber`, as suggested by [l0rinc](https://github.com/bitcoin/bitcoin/pull/30569#issuecomment-2279822735) and [ryanofsky](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1715587191)
- use the new `uint256::FromUserHex()` function for `-minimumchainwork`, [`-assumevalid`](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1711973562) and [`RANDOM_CTX_SEED`](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1711982050) parsing, minimizing the behaviour change in this PR and allowing 0x prefixes and too-short-input in all 3 cases. The main behaviour change now introduced by this PR is limited to more explicit handling of invalid hex characters and too-long-input, which I believe won't be controversial.
- remove the now unused `IsHexNumber()`. Most of its unit tests have been migrated to `uint256_tests/from_user_hex`.
- improved testing in `validation_chainstatemanager_tests`:
    - refactored `SetOptsFromArgs` to be templated and reusable, and allow for better error reporting by moving the BOOST_ calls to the edges as much as ergonomically possible
    - added unit tests for invalid-hex-char or too-many-char cases for -minimumchainwork and -assumevalid
- added and cleaned up some of the `uint256_tests/from_user_hex tests [1](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1725705844), [2](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1725701490), [e](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1725699517) 
- terminology: [consistently use "hex digits"](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1725732478) whenever the character is indeed a [0-9, a-f] value, and "(non-hex) character" when it isn't or we're not sure yet.

l0rinc commented:
I like it a lot more now, thanks for considering the feedbacks.

The only part that still bugs me a bit is the naming of `FromUserHex` -> the "user" is on a different abstraction level, I'd prefer mentioning the behavior change compared to the sibling method (i.e. it's more forgiving/leninent for the inputs), rather than the motivation for adding the method (the users sometimes prefer different formats).

ACK 81554aac80bf2270db977c110c37acc7e8034194

stickies-v commented:
> Sorry for not understanding that you had it in the works with my prior suggestion.

No problem at all, I decided to give this PR a bit of a rest to focus on reviewing and aligning with related PRs, and think about the suggestions made here, hence the delay in updating this. Will quickly address feedback again from here on.

> The only part that still bugs me a bit is the naming of `FromUserHex` -> the "user" is on a different abstraction level, I'd prefer mentioning the behavior change compared to the sibling method (i.e. it's more forgiving/leninent for the inputs), rather than the motivation for adding the method (the users sometimes prefer different formats).

I think it's a valid concern, and I agree that generally functions should be named in line with what they do, not with how they're used - and `FromHexLenient` would be a good name. The reason I'm sticking with `FromUserHex` is that it's 100% a convenience function that we've implemented specifically for dealing with user input, and the name makes that very clear imo. I'd also prefer to avoid bikeshedding over this, since it does touch quite a few LoC, commits, commit messages, PR descriptions, ... and I think both options are good.

Thank you both for your continued review!

ryanofsky commented:
Code review ACK 81554aac80bf2270db977c110c37acc7e8034194. Thanks for all the changes. This seems both simpler and more backward compatible now.

---

re: https://github.com/bitcoin/bitcoin/pull/30569#issue-2443189383

In PR description:

>* test: the optional `RANDOM_CTX_SEED` env var is now required to consist only of up to 64 hex digits (optionally prefixed with "0x"), otherwise the program will abort

I think it would be clearer and more consistent with the rest of the description if it said "test: the optional `RANDOM_CTX_SEED` env var will now cause tests to abort when it contains invalid hex characters, or when it is longer than 64 hex digits"

hodlinator commented on test/functional/feature_minchainwork.py:
```diff
@@ -23,7 +23,6 @@
 
 # 2 hashes required per regtest block (with no difficulty adjustment)
 REGTEST_WORK_PER_BLOCK = 2
-
```
nit: Inadvertent newline removal?

stickies-v replied:
yeah, I could've sworn a previous version introduced this newline and this was me reverting it, but I suppose it must be a rebase artifact, thanks for spotting and fixed now.

hodlinator reviewed (approved):
re-ACK 81554aac80bf2270db977c110c37acc7e8034194

`git range-diff master cf88a55e97719aabd62f0b608df0800fef8304de 81554aac80bf2270db977c110c37acc7e8034194`

Only changes `BOOST_AUTO_TEST_CASE(from_user_hex)` tests, error strings, commit messages and revives the `LogPrintf` -> `LogInfo` update.

stickies-v commented on src/test/util/random.cpp:
```diff
@@ -7,32 +7,39 @@
+            if (auto num_parsed{uint256::FromUserHex(num)}) {
+                return *num_parsed;
+            } else {
+                std::cerr << RANDOM_CTX_SEED << " must consist only of up to " << uint256::size() * 2 << " hex digits (\"0x\" prefix allowed), was set to: '" << num << "'.\n";
```
Thanks, taken.

stickies-v commented on src/test/uint256_tests.cpp:
```diff
@@ -386,6 +386,34 @@ BOOST_AUTO_TEST_CASE(from_hex)
+    BOOST_CHECK_EQUAL(uint256::FromUserHex(valid_hex_65.substr(2, 64)).value().ToString(), valid_hex_65.substr(2, 64));
+    BOOST_CHECK_EQUAL(uint256::FromUserHex(valid_hex_65.substr(0, 66)).value().ToString(), valid_hex_65.substr(2, 64));
+
+    BOOST_CHECK(!uint256::FromUserHex("0x0 "));                 // no spaces at end,
```
Done, and reorganized the test a bit to use `std::string valid_hex_64` instead of `std::string_view valid_hex_65` which makes cat operations significantly less verbose, and makes the test more readable imo.

stickies-v commented on src/test/uint256_tests.cpp:
```diff
@@ -386,6 +386,34 @@ BOOST_AUTO_TEST_CASE(from_hex)
+    constexpr std::string_view valid_hex_65{"0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0"}; // 0x prefix and 65 hex chars
+    static_assert(valid_hex_65.size() == 2 + 65);
+    BOOST_CHECK_EQUAL(uint256::FromUserHex(valid_hex_65.substr(2, 64)).value().ToString(), valid_hex_65.substr(2, 64));
+    BOOST_CHECK_EQUAL(uint256::FromUserHex(valid_hex_65.substr(0, 66)).value().ToString(), valid_hex_65.substr(2, 64));
```
Resolved by [replacing](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1727507122) `valid_hex_65` with `valid_hex_64` which I think addresses the confusion, too?

stickies-v commented on src/test/uint256_tests.cpp:
```diff
@@ -386,6 +386,34 @@ BOOST_AUTO_TEST_CASE(from_hex)
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0x10").value(), uint256{0x10});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("10").value(), uint256{0x10});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0xff").value(), uint256{0xff});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("ff").value(), uint256{0xff});
```
Hmm, fair point. I think I dropped it because of #30560 dropping mixed case support for the consteval ctor, but that's unrelated to this pull of course so keeping the test is the right approach. I've added mixed case into 4 existing checks, but I'd still like to drop the `Ffa` tests since we can't use the `uint8_t` ctor which would make an equality test quite verbose for I think nothing that's not already tested in the other lines?

```cpp
    BOOST_CHECK_EQUAL(uint256::FromUserHex("0xFf").value(), uint256{0xff});
    BOOST_CHECK_EQUAL(uint256::FromUserHex("Ff").value(), uint256{0xff});
    const std::string valid_hex_64{"0x0123456789abcdef0123456789abcdef0123456789ABDCEF0123456789ABCDEF"};
    BOOST_REQUIRE_EQUAL(valid_hex_64.size(), 2 + 64); // 0x prefix and 64 hex digits
    BOOST_CHECK_EQUAL(uint256::FromUserHex(valid_hex_64.substr(2)).value().ToString(), ToLower(valid_hex_64.substr(2)));
    BOOST_CHECK_EQUAL(uint256::FromUserHex(valid_hex_64.substr(0)).value().ToString(), ToLower(valid_hex_64.substr(2)));
```

stickies-v reviewed (commented):
Force-pushed to address all review comments, limited to nits and tests so should be an easy re-review:
- [grammar nit](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1726880388) for `RANDOM_CTX_SEED`
- [added](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1727507122) `from_user_hex` test case to cover more > 64 char cases and cleaned up the test a bit by replacing `valid_hex_65` with `valid_hex_64` which should be less confusing
- [re-added mixed case testing](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1727569784) in `from_user_hex`
- [undid](https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1727696360) unintentional newline removal
> I think it would be clearer and more consistent with the rest of the description if it said "test: the optional RANDOM_CTX_SEED env var will now cause tests to abort when it contains invalid hex characters, or when it is longer than 64 hex digits"

Thanks, PR description updated with your suggestion.

hodlinator commented:
re-ACK 18d65d27726bf9fc7629b8e794047a10c9cf6156

`git range-diff master 81554aac80bf2270db977c110c37acc7e8034194 18d65d27726bf9fc7629b8e794047a10c9cf6156`

Changes: Grammar, added&adjusted tests (passed), undid unintentional new line.

ryanofsky reviewed (approved):
Code review ACK 18d65d27726bf9fc7629b8e794047a10c9cf6156. Very nice change that cleans up the API, adds checking for invalid values, makes parsing of values more consistent, and adds test coverage.

Just small test, error message, and comment changes since last review

l0rinc commented:
While I'm not a fan of mixed casing, I agree that it's probably outside the scope of current change.

ACK 18d65d27726bf9fc7629b8e794047a10c9cf6156

achow101 commented:
ACK 18d65d27726bf9fc7629b8e794047a10c9cf6156
